### DISCUSSION TITLE
MultipartFormDataOutput as an entity does not populate boundary in Content-Type header

### DISCUSSION THREAD

**Discussion by mqchen-ws** [resteasy/resteasy#4267] [General]
Version: 6.2.9.Final

Example code
```
    public static void main(String[] args) {
        final Client client = ResteasyClientBuilder.newClient();
        final var multipart = new MultipartFormDataOutput();
        multipart.addFormData("foo", "foo-value", MediaType.TEXT_PLAIN_TYPE);
        multipart.addFormData("bar", "bar-value", APPLICATION_OCTET_STREAM_TYPE, "bar.pdf");
        final var multipartEntity = Entity.entity(multipart, MULTIPART_FORM_DATA_TYPE);
        final Invocation invocation = client.target("localhost:8080/test")
                .request()
                .buildPost(multipartEntity);
        // breakpoint here + inspect invocation
        System.out.println(invocation);
    }
```
<img width="961" alt="image" src="https://github.com/user-attachments/assets/58d9d090-c796-4f06-918e-77ad80b23e0a">

The entity is of type `MultipartFormdataOutput` and it has a `boundary` parameter which looks like a randomly generated UUID. However, in `Invocation.headers` the `Content-Type` is set to only `multipart/form-data`, without the boundary.

Explicitly constructing the media type and setting the `boundary` as a parameter, instead of using `MULTIPART_FORM_DATA_TYPE`, works:
```
final var mediaType = new MediaType("multipart", "form-data", Map.of("boundary", multipart.getBoundary()));
```

Since the entity already knows the boundary, it seems like the `buildPost` should be able to set the boundary without explicitly constructing the media type?

**Comment by jamezp** [comment]
What version of RESTEasy are you using?

**Reply by mqchen-ws** [reply]
6.2.9.Final

**Reply by jamezp** [reply]
Sorry, I just noticed that it was the first thing in the question :facepalm:. I'll have a look at this.

**Reply by mqchen-ws** [reply]
No, you were right and I didn't have it there originally. I edited it for anyone looking at it in the future 😅

**Comment by jamezp** [comment]
@mqchen-ws Are you seeing some kind of an error? While the headers on the invocation may not have the boundary set, it's set in the request. If I dump the request headers I see something like:
```
Connection: "Keep-Alive"
User-Agent: "Apache-HttpClient/4.5.14 (Java/11.0.24)"
Host: "localhost:8081"
Content-Type: "multipart/form-data; boundary=e81e31c2-8239-4287-a17e-92c419a27dce"
Accept-Encoding: "gzip, deflate"
Content-Length: "484"
```

**Reply by mqchen-ws** [reply]
I found this while writing a test against my endpoint handling multipart requests. If I don't explicitly set the boundary, these are the headers I get from the `HttpServletRequest`:

```
Transfer-Encoding: chunked
traceparent: 00-e776f08e736c4f6a3f2dc7bc3880be91-7e314e0d71529590-00
User-Agent: service-test/unknown
Host: localhost:49681
Accept-Encoding: gzip
Content-Type: multipart/form-data
```

This is using the `JettyClientEngine` and an `asyncExecutor` (set with `ResteasyClientBuilder#executorService`) on the client.

**Reply by jamezp** [reply]
Interesting. I'll try to update my reproducer and see if I can't reproduce this.

**Reply by mqchen-ws** [reply]
We do also have a few other settings in building the client, but at a glance it didn't look _too_ significant. I'll try stripping some of it away and see if I can get a more minimal case as well.

**Reply by mqchen-ws** [reply]
Ok, looks like using the default `ApacheHttpClient` works fine on my end as well. Trying now with the jetty engine.

**Reply by mqchen-ws** [reply]
Ah! I simplified it down to this in my test:

```
        final var client = new ResteasyClientBuilderImpl()
//                .httpEngine(new JettyClientEngine(new HttpClient()))
                .build();
```

If that line is uncommented out and the jetty engine is used, I don't see the boundary on the header. But without that line, the `ApacheHttpClient` seems to do the right thing.

**Reply by jamezp** [reply]
Okay good to know. That sounds like a bug in the `JettyClientEngine` to me. I'll try and create a reproducer and test for it, then file an issue and get it fixed.

For my own curiosity, why use the jetty-client? I'm asking because I really want to change the default client to be something other than Apache HTTP Client 4.x and support http/2.

**Reply by mqchen-ws** [reply]
> For my own curiosity, why use the jetty-client? I'm asking because I really want to change the default client to be something other than Apache HTTP Client 4.x and support http/2.

That I am not entirely sure, it was set up this way before I started working on the code 😅. I can try asking around.

**Reply by stevenschlansker** [reply]
@jamezp , we initially went with Jetty since we'd already selected it for our server (good experiences at prior projects) and liked the "clean room" design that has good support for async (as well as sync) from the start. Jetty also has good extensions for websockets and http/2, although we haven't in the end used either much yet. We also had positive experiences with the maintainers over time. 

I'd had "okay" experiences with apache httpclient - it usually works at the end of the day, but really felt the "legacy" aspects at times.

I'm puzzled that multipart entities in particular don't work - maybe it needs special handling?

**Reply by jamezp** [reply]
Thank you @stevenschlansker. That is great information :)

I'm not too sure yet why the headers are behaving differently. I don't know if it's a RESTEasy issue or a Jetty Client issue. I'm leaning towards RESTEasy, but I've not yet looked at it in detail.

**Comment by jamezp** [comment]
Thank you for all the information on this. I've created https://issues.redhat.com/browse/RESTEASY-3532.

**Comment by jamezp** [comment]
This should fix it https://github.com/resteasy/resteasy/pull/4282. The issue was the headers were set before the content was written, which seems to be where he boundary is added to the header.
